matchmaker.js

```diff
@@ -0,0 +1,146 @@
+import Player from './player.js'
+import GameRoom from './gameroom.js';
+import RoomManager from './roommanager.js'
+import { WebSocketServer } from 'ws'
+import { v4 as uuidv4 } from 'uuid';
+
+import { upload_to_blob_storage } from './blobstorage.js';
+
+class Matchmaker {
+    constructor(room_manager) {
+        this.banned_characters = ["carmine"]
+        this.room_manager = room_manager
+        this.match_queues = {}
+
+        this.match_types = {
+            "untimed" : {
+              "starting_timer" : 0,
+              "enforce_timer" : false,
+              "minimum_time_per_choice" : 0},
+            "timed" : {
+              "starting_timer" : 12,
+              "enforce_timer" : true,
+              "minimum_time_per_choice" : 20},
+            "speed" : {
+              "starting_timer" : 6,
+              "enforce_timer" : true,
+              "minimum_time_per_choice" : 10}
+        }
+    
+    for (queue_type in this.match_types.keys) {
+            this.match_queues[queue_type] = {
+                "has_player_waiting" : false,
+                "waiting_player" : player = null,
+                "waiting_json_data" : object = null,
+                "waiting_version" : Number
+            }
+        }
+    }
+
+    remove_from_matchmaking(player) {
+        for (queue in this.match_queues.keys) {
+            if (queue["waiting_player"] == player) {
+                this.match_queues[queue]["has_player_waiting"] = false
```

Comment:
so, you don't really need an extra bool for has_player_waiting, you can just use whether waiting_player is null or not. Then you can set it to null here